Dongjo/week5 assignment 1

ParkDongJo opened the pull request:
이번주는 Top - Down 방식으로 TDD 적용중

ParkDongJo committed:
[week5-1] 기본 구조 구축하기

ParkDongJo committed:
[week5-2] App.test.jsx 테스트코드 작성하기

[week5-2] App test 코드 추가

ParkDongJo committed:
[week5-2] fixture 데이터 & mocks 함수추가

ParkDongJo committed:
[week5-2] App 컴포넌트 1차작성

ParkDongJo committed:
[week5-2] RegionsContainer.test 테스트 코드작성

ParkDongJo committed:
[week5-2] RegionsContainer 컴포넌트 구현

ParkDongJo committed:
[week5-2] Region.test 테스트 코드 작성

ParkDongJo committed:
[week5-2] Region 컴포넌트 구현하기

ParkDongJo committed:
[week5-2] RegionContainer 분리 반영 테스트코드 작성

ParkDongJo committed:
[week5-2] RegionContainer로 지역목록 분리시킴

ParkDongJo committed:
[week5-2] setRegions reducer 함수 테스트 코드작성

ParkDongJo committed:
[week5-2] reducer에 setRegions 함수 추가하기

ParkDongJo committed:
[week5-2] api fetch Regions 테스트 코드 작성

ParkDongJo committed:
[week5-2] api fetchRegions 코드작성

ParkDongJo committed:
[week5-2] App dispatch loadRegion 호출 테스트 코드 작성

ParkDongJo committed:
[week5-2] aciton 함수 작성 후 App에서 loadRegion() 호출하기

ParkDongJo committed:
[week5-2] Reducer 호출 안되는 버그 수정하기

ParkDongJo committed:
[week-5] RegionCotainer에 Region 컴포넌트 적용하기

ParkDongJo committed:
[week5-2] CtegoriesContainer.test 테스트 코드 작성

ParkDongJo committed:
[week5-2] CategoriesContainer.jsx 컴포넌트 구현 & App에 적용

ParkDongJo committed:
[week5-2] App.test 테스트 코드 에러 대응 테스트코드 작성

ParkDongJo committed:
[week5-2] Category.test 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] Category 컴포넌트 구현하기

ParkDongJo committed:
[week5-2] reducer category.test 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] reducer category에 setCategories() 대응 코드구현하기

ParkDongJo committed:
[week5-2] CategoriesContainer에 Category 컴포넌트 적용하기

ParkDongJo committed:
[week5-2] api.test fetchCategories() 테스트코드 작성하기

ParkDongJo committed:
[week5-2] api에 fetchCategories() 구현하기

ParkDongJo committed:
[week5-2] action에 loadCategories() 함수구현

ParkDongJo committed:
[week5-2] App.test 테스트코드 dispatch 호출 times 2번 호출 적용하기

ParkDongJo committed:
[week5-2] App 컴포넌트에 loadCategories() 호출 적용하기

ParkDongJo committed:
[week5-2] reducers 에 category reducer 추가시키기

ParkDongJo committed:
[week5-2] RestaurantsContainer.test 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] RestaurantsContainer 컴포넌트 구현하기

ParkDongJo committed:
[week5-2] Restaurant.test 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] Restaurant 컴포넌트 구현하기

ParkDongJo committed:
[week5-2] App에 RestaurantsContainer 테스트 코드 적용하기

ParkDongJo committed:
[week5-2] App에 RestaurantsContainer 컴포넌트 적용하기

ParkDongJo committed:
[week5-2] reducer restaurant.test setRestaurants() 함수 테스트 코드 작성하기

* actions/index 에 setRestaurants() 함수구현
* reducer.test setResaurants() 테스트 코드 작성

ParkDongJo committed:
[week5-2] reducer restaurant setRestaurant action 대응 코드 작성하기

ParkDongJo committed:
[week5-2] api.test에 fetchRestaurants() 함수 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] api에 fetchRestraunts() 함수 구현하기

ParkDongJo committed:
[week5-2] App.test에 dispach 호출 times 3번으로 수정하기

ParkDongJo committed:
[week5-2] App에 loadRestaurants() 함수 적용하기 & action 구현

* actions/index 에 loadRestaurants() 함수 구현
* App 컴포넌트 useEffect() 에 loadRestaurants() 함수 호출

ParkDongJo committed:
[week5-2] reducer에 selected 초기값 수정하기

ParkDongJo committed:
[week5-2] RestaurantsContainer에 Restaurant 컴포넌트 적용하기

Kihyun92 commented on src/components/App.jsx:
```diff
@@ -0,0 +1,31 @@
+export default function App() {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(loadRegions());
+    dispatch(loadCategories());
+    dispatch(loadRestaurants());
+  });
```
useEffect를 통해서 action들을 dispatch 해주셨네요~ 혹시 두번째 인자를 추가하지 않은 이유가 있을까요?
이렇게 처리할 경우 지금은 의존성이 없기 때문에 문제가 없지만, 가능하면 두번째 인자를 명시해주면 더 의도를 드러낼 수 있을 것 같아요~ 지금은 초기 렌더될때 한번만 dispatch 해주면 될 것 같아요. 두번째 인자로 빈 배열을 넣어주면 좋을 것 같습니다~

- https://ko.reactjs.org/docs/hooks-reference.html#useeffect

Kihyun92 commented on src/containers/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,27 @@
+
+  return (
+    <ul>
+      {categories.map((category) => (
+        <Category
+          key={category.id}
+          name={category.name}
+          onClick={handleClick}
+        />
```
```suggestion
      {categories.map(({ id, name }) => (
        <Category
          key={id}
          name={name}
          onClick={handleClick}
        />
```

이 부분은 구조 분해 할당으로 처리해줘도 의도가 전달 될 것 같네요~

Kihyun92 commented on src/index.jsx:
```diff
@@ -1,9 +1,21 @@
 
 ReactDOM.render(
-  <App />,
+  (
+    <Provider store={store}>
+      <App />
+    </Provider>
+  ),
```
감싸고 있는 소괄호는 없어도 될것 같아요~

Kihyun92 commented on src/reducers/category.js:
```diff
@@ -0,0 +1,17 @@
+};
+
+const category = (state = initialState, action) => {
+  if (action.type === 'setCategories') {
```
if 문을 사용하지 않고 처리해볼 수 있을까요~?

`Object literal lookup`등 여러 방법이 있을 것 같아요~

- https://ultimatecourses.com/blog/deprecating-the-switch-statement-for-object-literals

Kihyun92 reviewed (changes requested):
리뷰가 조금 늦었네요 🙏 
`Red-Green-Refactoring` 사이클을 지키려고 노력하신 것 같아요 👍 
그리고 폴더 구조를 분리해서 관심사를 분리해보려 하신것 같아요 좋네요~

나머지 기능들도 차근차근 추가해보시면 금방 완성하실 수 있을 것 같습니다~

Kihyun92 commented:
![스크린샷 2020-10-24 오전 11 50 44](https://user-images.githubusercontent.com/35797540/97067503-c9676000-15f8-11eb-9ac1-7fc4b943620d.png)

추가로 질문 주셨던 컴포넌트 내부에서 사용되는 함수에 대한 테스트의 경우에는 외부로 빼는 대신, side effect가 없도록 순수 함수로 만들어야 할 것 같아요. 근데 사실 이 함수는 결국 selectedRegionName이 특정 지역일 때 체크 표시해서 UI에 보여주려는 것 같은데, UI에 대한 테스트를 해도 될 것 같구요~

말이 좀 복잡한 것 같은데 결국 근본적으로 이 함수를 왜 테스트하려는지 더 고민해보시면 좋을 것 같아요 👍

ParkDongJo committed:
[week5-2] api.test fetchRestaurants() 테스트 코드 리펙토링

* 선택된 regionName, categoryId 의 값 존재유무에 따라서 테스트코드 분리

ParkDongJo committed:
[week5-2] reducer region에 updateRegionName() 테스트코드 작성하기

ParkDongJo committed:
[week5-2] reducer region updateRegionName() 함수 작성하기

ParkDongJo committed:
[week5-2] RegionsContainer.test 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] RegionsContainer에 dispatch regionName 기능 구현하기

ParkDongJo committed:
[week5-2] Region.test 선택된 regionName 경우의 수에 따른 테스트 코드 분리하기

* 선택된 region name 이 있고 없고 경우의 수에 따른 테스트 코드 작성하기
* test() 에서 두가지 경우의 수에 대비한 describe - it 으로 구성하기₩

ParkDongJo committed:
[week5-2] Region에 isSelected props 추가하기

ParkDongJo committed:
[week5-2][실수] updateSelectedRegionName() 빠져먹은 commit 내용

[주의]
실수로 빠져먹은 코드내용을 뒤늦게 올린 커밋입니다. revert시 주의해주세요.
만약 revert 하고자 하신다면, 66cae0a31626afe790af94481aa360e3312ed135 커밋까지
revert 해주셔야 합니다

다만, 이때 이와 관련된 해당 커밋과 66cae0a31626afe790af94481aa360e3312ed135 커밋 사이의
다른 작업들도 모두 사라집니다

ParkDongJo committed:
[week5-2] category reducer 에 updateSelectedCategoryId() 테스트 코드 작성하기

ParkDongJo committed:
[week5-2] category reducer에 updateSelectedCategoryId() 대응 코드 작성하기

ParkDongJo committed:
[week5-2] CategoriesContainer.test 에 selectedCategoryId dispatch 테스트코드 추가

ParkDongJo committed:
[week5-2] CategoriesContainer에 dispatch selectedCategoryId 기능 구현하기

ParkDongJo committed:
[week5-2] Category.test에 isSelected props 테스트 코드 추가하기

ParkDongJo committed:
[week5-2] Category.test 선택된 CategoryId 경우의 수에 따른 테스트 코드작성

ParkDongJo committed:
[week5-2][개선필요] App.test 에 restaurant 초기값에 대한 코드수정

이후 test 코드에 대한 개선이 필요해보임, 현재로는 고정값에 의한
useSelector() 에서 state 값을 초기화 시키지만, API 결과값에 따른 초기화 코드로
개선이 필요함

ParkDongJo committed:
[week5-2] CategoriesContainer ui check표시 버그 수정

ParkDongJo committed:
[week5-1][코드고민] selected categoryId에 대한 테스트코드 고민 주석처리

ParkDongJo committed:
[week5-2] 상태변경 시 loadRestaurants() 호출 코드작성하기

ParkDongJo committed:
[week5-3][리펙토링] reducer category 테스트 코드 커버리지 높이기

ParkDongJo committed:
[week5][리펙토링] reducer category defualt 호출 시 테스트 통과 코드작성

ParkDongJo committed:
[week5-3][리펙토링] reducer region 테스트 코드 커버리지 높이기

ParkDongJo committed:
[week5][리펙토링] reducer region defualt 호출 시 테스트 통과 코드작성

ParkDongJo committed:
[week5][리펙토링] reducer category defualt 호출 시 테스트 통과 코드작성

ParkDongJo committed:
[week5-3][리펙토링] reducer restaurant 테스트 코드 커버리지 높이기

ParkDongJo committed:
[week5][리펙토링] reducer restaurant defualt 호출 시 테스트 통과 코드작성

ParkDongJo committed:
[week5-3] actions 에 코드 정리

Kihyun92 commented:
과제를 진행하면서 정말 많은 고민을 해보시는 것 같아 좋네요 👍  이미 보셨을 수도 있지만, action 테스트를 위해 `redux-mock-store`를 참고해보시면 좋을 것 같습니다~

- https://github.com/reduxjs/redux-mock-store
- https://jestjs.io/docs/en/mock-function-api#mockfnmockresolvedvaluevalue
